### PR TITLE
[stm32] Add support for STM32L0

### PR DESCRIPTION
**Pull request by chris-durand**
This adds support for STM32L0 devices.

- [x] Fix [USART data in modm-devices](https://github.com/modm-io/modm-devices/pull/58)
- [x] Add RCC for L0
- [x] Add Nucleo L031K6 BSP
- [x] Add Nucleo L031K6 blink example
- [x] Test example in hardware
- [x] Add L0 to CI jobs
- [x] Fix compilation of L0 timers
- [x] Fix UART interrupt loop for buffered tx / unbuffered rx
- [x] ~~Fix G0 usart shared interrupt problem due to updating device files~~ (fix postponed, device are ignored for now)

### CODE REVIEW

**Comment by chris-durand** [comment]
`synchronize_docs.py` crashes inside the uart `module.lb`, I will look into it tomorrow.
```
Traceback (most recent call last):
  File "/home/chris/modm/src/modm/platform/uart/stm32/module.lb", line 101, in prepare
    props["shared_irq_ids"] = range(int(parts[0]), int(parts[1]) + 1)
ValueError: invalid literal for int() with base 10: 'LPUART2'
```

It is caused by G0 shared irq ids not matching the format the lbuild file expects.

**Comment by salkinium** [comment]
Yup, there are more issues with shared IRQs with I2C and SPI. See #566. The fix is not trivial, I would recommend upgrading modm-devices (and STM32 CMSIS headers) and adding the G0Cx and G0Bx devices to ignored.txt. Then we can fix the problems in a separate PR.

Alternatively we first fix/test #566 and then continue with this PR. Depends on how quick you need L0 support in upstream modm (and if you want to fix the I2C/SPI shared IRQ problems).

**Comment by salkinium** [comment]
You also need to update the CMSIS header files to be in sync with modm-devices.

**Review by salkinium — APPROVED** [review]
Very nice, I like it (so far).

**Comment by chris-durand** [comment]
> Alternatively we first fix/test #566 and then continue with this PR. Depends on how quick you need L0 support in upstream modm (and if you want to fix the I2C/SPI shared IRQ problems).

I will put the failing G0 controllers in the ignore list for now since a colleague of mine at work needs L0 support soon.
The G0 problems are best handled in a separate PR.

**Comment by chris-durand** [comment]
I [fixed](https://github.com/modm-io/modm/pull/602/commits/cea91b2208e03e417b9187562ad33cf91b359efb) a problem in the UART driver that occurs when there is a tx buffer but no rx buffer. The controller got stuck on reception of data because the RX interrupt was enabled but the flag was never acknowledged.

**Comment by chris-durand** [comment]
Is there any documentation that needs a manual update after adding the L0 series?

**Comment by chris-durand** [comment]
@salkinium The files generated by synchronize_docs.py in the CI and my computer don't agree. I get from the CI:
```
Please synchronize the modm documentation:

    $ python3 tools/scripts/synchronize_docs.py

and then commit the results!
Error: Process completed with exit code 1.
```

**Comment by chris-durand** [comment]
More G0 issues (stm32g051k8u6, stm32g061k8u6, stm32g050k8t6):
```
-    stm32g061k8u6           0.5s

ERROR: In 'modm:cmsis:device':

Traceback (most recent call last):
  File "/__w/modm/modm/ext/st/module.lb", line 84, in common_rcc_map
    define = getDefineForDevice(device.identifier, match)
  File "/__w/modm/modm/ext/st/module.lb", line 35, in getDefineForDevice
    minlen = min(len(d) for d in deviceDefines)
ValueError: min() arg is an empty sequence
```

**Comment by chris-durand** [comment]
The G0{5, 6}*  issue should be fixed by updating the ST CMSIS submodule.

**Comment by salkinium** [comment]
> I fixed a problem in the UART driver that occurs when there is a tx buffer but no rx buffer. 

😱😱😱😱😱😱😱😱 That's quite horrifying…

> The G0{5, 6}* issue should be fixed by updating the ST CMSIS submodule.

You also need to add [these header patches](https://github.com/modm-io/modm/pull/566/commits/fca37794055a25d5ffaf2d0657c08dbab00c8002#diff-61f737daa7c65e54fc701fdac297a2e622d15785b3755a80820d8d27986b9bbc) (ST decided to remove these definitions, but only for G0 wtf).

**Comment by chris-durand** [comment]
> You also need to add these header patches (ST decided to remove these definitions, but only for G0 wtf).

I cherry-picked your commit and removed mine that updates the CMSIS headers. Let's see what the CI says.

**Comment by chris-durand** [comment]
@salkinium There is something weird going on with synchronize_docs.py. The output on my system is different to the CI. Both are running Ubuntu 20.04.

File: src/modm/platform/clock/stm32/rcc.cpp.in [hunk 1]
```diff
@@ -349,6 +352,21 @@ modm::platform::Rcc::enableOverdriveMode(uint32_t waitCycles)
...
+{
+	const uint32_t enableFlag = divideBy4 ? RCC_CR_HSIDIVEN : 0;
+	RCC->CR = (RCC->CR & ~RCC_CR_HSIDIVEN) | enableFlag;
+
+	while (isHsiPredivider4Active() != divideBy4) {
+		if (waitCycles-- == 0) {
+			return false;
+		}
+	}
+	return false;
```
Comment (chris-durand) [inline src/modm/platform/clock/stm32/rcc.cpp.in]:
```suggestion
	return true;
```
Oh no, I have to torture the CI again...

File: .github/workflows/compile-all.yml [hunk 2]
```diff
@@ -225,6 +225,26 @@ jobs:
...
+    runs-on: ubuntu-20.04
+    container:
+      image: ghcr.io/modm-ext/modm-build-cortex-m:latest
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v2
+      - name: Checkout code and update modm tools
+        run: |
+          (git submodule sync && git submodule update --init --jobs 8) & pip3 install --upgrade --upgrade-strategy=eager modm & wait
+      - name: Compile HAL for all STM32L1
```
Comment (chris-durand) [inline .github/workflows/compile-all.yml]: Copy paste error in description

**Comment by salkinium** [comment]
> The output on my system is different to the CI. Both are running Ubuntu 20.04.

I have the same problem… I'll check what I can do.

**Comment by chris-durand** [comment]
May I force-push to fix the two typos?

**Comment by salkinium** [comment]
Absolutely, it's your branch.

**Comment by chris-durand** [comment]
Thanks, will also split the `stm32l0-compile-all` tests into 2 parts. It was 33 minutes on the last run.

**Comment by chris-durand** [comment]
The split `stm32l0-compile-all-1` `stm32l0-compile-all-2` checks don't appear in the list. Does github only update the hal tasks when you trigger them by tagging?

**Comment by chris-durand** [comment]
That worked, thanks.

**Comment by salkinium** [comment]
Ideally I'd want a button that the reviewers but also the PR owner can press. But I guess we cannot have nice things…

**Comment by chris-durand** [comment]
> But I guess we cannot have nice things…

We are doing embedded on microcontrollers here. Since when are we the people who can have nice things :wink:

**Comment by salkinium** [comment]
The compile-all workflow now has 20 jobs, which is the parallism limit from GHA, so it'll block all progress until it has finished running…

**Comment by chris-durand** [comment]
It seems https://github.com/modm-io/modm-devices/pull/58 is ready for merging.

**Comment by salkinium** [comment]
I still cannot fix the dox sync problem. I'm going to do the release 2021q1 now, and I'm cherry-picking the UART fix, since that's really important.

**Review by salkinium — APPROVED** [review]
Ok, I approve the L0 releated changed, so thanks!

However, the CI is still broken, so I'm going to rebase and force push your branch some. Sorry for the emails…

**Comment by chris-durand** [comment]
Should I look into finding out why the CI is broken or will you? I am working on different stuff right now, but I might get to it later. My idea would be to get the build docker container and debug it in there.

**Comment by salkinium** [comment]
> Should I look into finding out why the CI is broken or will you? 

No, I will, hence a few commits. I suspect some issues with Python 3.8 (CI) vs 3.9 (local), but I'm not sure yet.

**Comment by chris-durand** [comment]
I have 3.8 as well as I am running the same OS version (Ubuntu 20.04) on the machine I am writing this from. My suspicion is right now some python package version.

**Comment by chris-durand** [comment]
The `build-docs-test` has a broken pip command: `pip3 --upgrade` instead of `pip3 install --upgrade`. Don't know if that is related to the problem.

**Comment by salkinium** [comment]
I rebased onto develop and now the script does indeed return the right thing, but now the CI is broken…

**Comment by chris-durand** [comment]
Docker pull gets an HTTP 500 from ghcr.io:
```
  /usr/bin/docker pull ghcr.io/modm-ext/modm-build-cortex-m:latest
  Error response from daemon: Head https://ghcr.io/v2/modm-ext/modm-build-cortex-m/manifests/latest: received unexpected HTTP status: 500 Internal Server Error
  Warning: Docker pull failed with exit code 1, back off 7.528 seconds before retry.
  ```

File: src/modm/board/nucleo_l031k6/board.hpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+	static constexpr uint32_t Spi1   = Apb2;
+
+	static constexpr uint32_t Usart2 = Apb1;
+
+	static constexpr uint32_t I2c1   = Apb1;
+
+	static constexpr uint32_t Timer2  = Apb1Timer;
+	static constexpr uint32_t Timer21 = Apb2Timer;
+	static constexpr uint32_t Timer22 = Apb2Timer;
```
Comment (rleh) [inline src/modm/board/nucleo_l031k6/board.hpp]: Most BSP have this entries alphabetically ordered.

You could add the Comp1 and Comp2 entries (= Apb2) if someone is going to enable/port the comparator driver for L0 in the future...

grafik

**Review by rleh — APPROVED** [review]
Nice work!

**Comment by rleh** [comment]
> but now the CI is broken…

> Docker pull gets an HTTP 500 from ghcr.io: [...]

https://www.githubstatus.com/:
grafik
I guess we have to wait...

File: src/modm/board/nucleo_l031k6/board.hpp (hunk 3 above)
Comment (chris-durand) [reply src/modm/board/nucleo_l031k6/board.hpp]: Thanks, now alphabetically ordered and with comparator clock.

**Comment by chris-durand** [comment]
Now even the arm.com download server fails in travis ci. Microsoft cloud has network issues: https://status.azure.com/en-us/status.

**Comment by salkinium** [comment]
Yay, let's merge this!

**Pull request merged**